python/lsst/daf/butler/core/schema.py

```diff
@@ -83,317 +116,341 @@ def process_result_value(self, value, dialect):
+    def addForeignKeys(self, tableName: str, schema: Schema):
+        """Add SQLAlchemy foreign key constraints for this table to the
+        corresponding entry in the given schema.
 
-            May also contain:
+        Parameters
+        ----------
+        tableName : `str`
+            Name of the logical table.
+        schema : `Schema`
+            Object represening the full schema.  Will be modified in-place.
+
+        Notes
+        -----
+        This must be called after `toSqlAlchemy` has been called on all tables
+        in the schema.
+        """
+        table = schema.tables[tableName]
+        if not self.isView():
+            for foreignKeySpec in self.foreignKeys:
+                if foreignKeySpec.table not in schema.views:
+                    table.append_constraint(foreignKeySpec.toSqlAlchemy(tableName, schema.metadata))
 
-            - onDelete, one of "SET NULL" or "CASCADE".
 
-        Returns
-        -------
-        src : `tuple`
-            Sequence of field names in the local table.
-        tgt : `tuple`
-            Sequence of table-qualified field names in the remote table.
-        tgtTable : `str`
-            Name of the target table.
-        onDelete : `str`, optional
-            One of "SET NULL", "CASCADE", or `None`.
-        """
-        src = tuple(iterable(constraintDescription["src"]))
-        tgt = tuple(iterable(constraintDescription["tgt"]))
-        tgtTable, _ = tgt[0].split(".")
-        assert all(t.split(".")[0] == tgtTable for t in tgt[1:])
-        onDelete = constraintDescription.get("onDelete", None)
-        return src, tgt, tgtTable, onDelete
+class SchemaConfig(ConfigSubset):
+    component = "schema"
+    requiredKeys = ("version", "tables")
+    defaultConfigFile = "schema.yaml"
+
+    def toSpec(self) -> Dict[str, TableSpec]:
+        return {name: TableSpec.fromConfig(config) for name, config in self["tables"].items()}
+
+
+class Schema:
+    """The SQL schema for a Butler Registry.
+
+    Parameters
+    ----------
+    spec : `dict`, optional
+        Dictionary mapping `str` name to `TableSpec`, as returned by
+        `SchemaConfig.toSpec`.  Will be constructed from default configuration
+        if not provided.
+
+    Attributes
+    ----------
+    metadata : `sqlalchemy.MetaData`
+        The sqlalchemy schema description.
+    tables : `dict`
+        A mapping from table or view name to the associated SQLAlchemy object.
+        Note that this contains both true tables and views.
+    views : `set`
+        The names of entries in ``tables`` that are actually implemented as
+        views.
+    """
+    def __init__(self, spec=None):
+        if spec is None:
+            config = SchemaConfig()
+            spec = config.toSpec()
+        self.metadata = MetaData()
+        self.views = set()
+        self.tables = dict()
+        for tableName, tableSpec in spec.items():
+            tableSpec.toSqlAlchemy(tableName, self)
```

Comment:
On my next big refactor ticket I'd like to just remove this `Schema` class and then start using the return values; now that it isn't responsible for actually translating `Config` to SQLAlchemy anymore, `Schema` is not much more than a dictionary.  I've kept it around for now because switching to a plain `dict` would require more changes to `SqlRegistry` that are otherwise unrelated to this ticket.  In the meantime, I'll add a code comment.